src/main/java/com/magicvet/service/InputValidator.java

```diff
@@ -0,0 +1,32 @@
+package main.java.com.magicvet.service;
+
+import main.java.com.magicvet.Main;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class InputValidator {
+    public static String validateInputForPattern(String input, String patternString, String validInput) {
+        Pattern pattern = Pattern.compile(patternString);
+        Matcher matcher = pattern.matcher(prepareInputForMatch(input,patternString));
+        while (!matcher.matches()){
+            System.out.println("The input '" + input + "' is not valid! Valid input is: " + validInput);
+            System.out.print("Please try again: ");
+            input = Main.SCANNER.nextLine();
+            matcher = pattern.matcher(prepareInputForMatch(input,patternString));
+        }
+        return prepareInputForMatch(input,patternString);
+    }
```

Comment:
Combining the functionality of prepareInputForMatch directly into validateInputForPattern may be simpler and improve code readability. This consolidation will also eliminate the need for a separate method call and reduce code complexity.
```java
	 matcher = pattern.matcher(input);
        }
        // Налаштування регістру
        if (patternString.equals(patternString.toLowerCase())) {
            return input.trim().toLowerCase();
        } else if (patternString.equals(patternString.toUpperCase())) {
            return input.trim().toUpperCase();
        } else {
            return input.trim();
        }
```

Reply:
This won't work for a patternString containing regex metacharacters. See example above.
I'm also not happy with this code. May be it would be better to skip it at all and don't let the user case-insensitive input.
But let it be for a while until the next improvement )